flang/lib/Lower/IO.cpp

```diff
@@ -964,15 +964,43 @@ getFormat<Fortran::parser::PrintStmt>(
                    strTy, lenTy, labelMap, assignMap);
 }
 
-static std::tuple<mlir::Value, mlir::Value, mlir::Value>
+static std::tuple<mlir::Value, mlir::Value>
+genBufferFromExpr(Fortran::lower::AbstractConverter &converter,
+                  mlir::Location loc, const SomeExpr *e) {
+  auto &builder = converter.getFirOpBuilder();
+  if (Fortran::semantics::ExprHasTypeCategory(
+          *e, Fortran::common::TypeCategory::Character)) {
+    // Helper to query [BUFFER, LEN].
+    Fortran::lower::CharacterExprHelper helper(builder, loc);
+    auto dataLen = helper.materializeCharacter(converter.genExprValue(*e));
+    return {dataLen.first, dataLen.second};
+  }
+}
+
+/// Generate a reference to a buffer and the length of buffer.
+/// There are 3 cases - An IoUnit can be variable or
+/// a ScalarIntExpr (i.e FileUnitNumber) or a *.
+static std::tuple<mlir::Value, mlir::Value>
 genBuffer(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
           const Fortran::parser::IoUnit &iounit, mlir::Type strTy,
           mlir::Type lenTy) {
-  [[maybe_unused]] auto &var = std::get<Fortran::parser::Variable>(iounit.u);
-  TODO();
+  // Variable
+  if (auto *var = std::get_if<Fortran::parser::Variable>(&iounit.u)) {
+    auto e = Fortran::semantics::GetExpr(*var);
+    return genBufferFromExpr(converter, loc, e);
+  } else if (auto *fileUnit =
+                 std::get_if<Fortran::parser::FileUnitNumber>(&iounit.u)) {
+    // FileUnitNumber
+    auto e = Fortran::semantics::GetExpr(*fileUnit);
+    return genBufferFromExpr(converter, loc, e);
+  } else {
+    // Star
+    assert(std::get_if<Fortran::parser::Star>(&iounit.u));
+    TODO();
+  }
 }
 template <typename A>
-std::tuple<mlir::Value, mlir::Value, mlir::Value>
+std::tuple<mlir::Value, mlir::Value>
```

Comment:
+1

Reply:
std::pair could be used here, of course. But a std::tuple is perfectly acceptable.